webdictionary/src/main/java/com/eis0/kademlia/KademliaId.java

```diff
@@ -146,7 +181,48 @@ public KademliaId xor(KademliaId nid) {
         return new KademliaId(result);
     }
 
+    /**
+     * Generates a NodeId that is some distance away from this NodeId
+     *
+     * @param distance The index of the first bit that should be different,
+     *                 From 0 to 159, where 0 changes the most significant bit
+     * @return The newly generated NodeId with a given distance from this.
+     * @throws IllegalArgumentException if the distance is < 0 or >= 160
+     * @author Edordo Raimondi, improved by Marco Cognolato
+     */
+    public KademliaId generateNodeIdByDistance(int distance) {
+        if(distance < 0 || distance >= ID_LENGTH) throw new IllegalArgumentException();
+
+        byte[] result = keyBytes.clone();
+
+        // calculate the index of the byte and bit to update
+        int byteToUpdateIndex = (distance) / 8;
+        int bitToUpdateIndex = 7 - (distance % 8);
+
+        //change only the bit at the distance requested
+        result[byteToUpdateIndex] = (byte)(result[byteToUpdateIndex] ^ 0x01<<bitToUpdateIndex);
+        return new KademliaId(result);
+    }
 
+    /**
+     * Returns the index of the first bit set to 1, starting from the
+     * most significant digit
+     *
+     * @return Index of the first bit set to 1, returns -1 if there's no bit set to 1
+     * @author EdoardoRaimondi, Marco Cognolato
+     */
+    public int getFirstSetBitIndex() {
+        for(int byteIndex = 0; byteIndex < keyBytes.length; byteIndex++) {
+            byte currentByte = keyBytes[byteIndex];
+            //8 bits in a byte, from 0 to 7
+            for(int bitIndex = 7; bitIndex >= 0; bitIndex--) {
+                if(((0x01 << bitIndex) & currentByte) == (0x01 << bitIndex)){
+                    return (7-bitIndex) + byteIndex * 8;
+                }
+            }
+        }
+        return -1;
+    }
 
     /**
      * Gets the bit distance from this NodeId to another NodeId
```

Comment:
I'll explain rows 258 and 259, even though I didn't work on those.
First we create a BigInteger from the byte array (the 1 is a signum value used by the constructor to say how the number should behave. It can be -1 meaning the number is negative, 1 meaning the number is positive, and 0 meaning the whole number is zero)
Then we srite that number as a string, formatting it in the process.
A string formatted for hex numbers is in the form "%0nX", where n is the minimum number of characters in hex form.
so for example in the format "%05X" you can write "0xFFFFF" as well as "0x000AB" since there are 5 characters after the "x"
What we do is, since 1 byte is 2 hex characters, we take the length of the array and "multiply by two", which you can also do if you bitshift 1 time to the left.
Again, I didn't work on that so I don't know why that's done that way.

Reply:
Regarding rows 269-275 I think I can just remove all that method and simply call `hexRepresentation` since I believe they're doing the same thing in different ways.

Reply:
Very nice to know! Thank you for the explanation.